libs/megadrive-alloc/src/heap.rs

```diff
@@ -0,0 +1,139 @@
+    }
+
+    /// Allocates a chunk of the given size with the given alignment. Returns a pointer to the
+    /// beginning of that chunk if it was successful. Else it returns `None`.
+    /// This function scans the list of free memory blocks and uses the first block that is big
+    /// enough. The runtime is in O(n) where n is the number of free blocks, but it should be
+    /// reasonably fast for small allocations.
+    pub fn allocate_first_fit(&mut self, layout: Layout) -> Result<*mut u8, AllocError> {
+        let mut size = layout.size();
+        if size < HoleList::min_size() {
+            size = HoleList::min_size();
+        }
+        let size = align_up(size, mem::align_of::<Hole>());
+        let layout = Layout::from_size_align(size, layout.align()).unwrap();
+
+        self.holes.allocate_first_fit(layout)
+    }
+
+    /// Frees the given allocation. `ptr` must be a pointer returned
+    /// by a call to the `allocate_first_fit` function with identical size and alignment. Undefined
+    /// behavior may occur for invalid arguments, thus this function is unsafe.
+    ///
+    /// This function walks the list of free memory blocks and inserts the freed block at the
+    /// correct place. If the freed block is adjacent to another free block, the blocks are merged
+    /// again. This operation is in `O(n)` since the list needs to be sorted by address.
+    pub unsafe fn deallocate(&mut self, ptr: *mut u8, layout: Layout) {
+        let mut size = layout.size();
+        if size < HoleList::min_size() {
+            size = HoleList::min_size();
+        }
+        let size = align_up(size, mem::align_of::<Hole>());
+        let layout = Layout::from_size_align(size, layout.align()).unwrap();
+
+        self.holes.deallocate(ptr, layout);
+    }
+
+    /// Returns the bottom address of the heap.
+    pub fn bottom(&self) -> usize {
+        self.bottom
+    }
+
+    /// Returns the size of the heap.
+    pub fn size(&self) -> usize {
+        self.size
+    }
+
+    /// Return the top address of the heap
+    pub fn top(&self) -> usize {
+        self.bottom + self.size
+    }
+
+    /// Extends the size of the heap by creating a new hole at the end
+    ///
+    /// # Unsafety
+    ///
+    /// The new extended area must be valid
+    pub unsafe fn extend(&mut self, by: usize) {
+        let top = self.top();
+        let layout = Layout::from_size_align(by, 1).unwrap();
+        self.holes.deallocate(top as *mut u8, layout);
+        self.size += by;
+    }
+}
+
+/// Align downwards. Returns the greatest x with alignment `align`
+/// so that x <= addr. The alignment must be a power of 2.
+pub fn align_down(addr: usize, align: usize) -> usize {
+    if align.is_power_of_two() {
+        addr & !(align - 1)
+    } else if align == 0 {
+        addr
+    } else {
+        panic!("`align` must be a power of 2");
+    }
+}
+
+const HEAP: RefCell<Heap> = RefCell::new(Heap::empty());
```

Comment:
```suggestion
static HEAP: RefCell<Heap> = RefCell::new(Heap::empty());
```

We expect this to be mutated, so it shouldn't be const.